pkg/distributor/distributor.go

```diff
@@ -1028,6 +1037,61 @@ func (d *Distributor) prePushForwardingMiddleware(next push.Func) push.Func {
 	}
 }
 
+// prePushEphemeralMiddleware is used as push.Func middleware in front of push method.
+// If marking series as ephemeral is enabled, this middleware uses the ephemeral series
+// provider to determine whether a time series should be marked as ephemeral.
+func (d *Distributor) prePushEphemeralMiddleware(next push.Func) push.Func {
+	if !d.cfg.EphemeralMetricsEnabled {
+		return next
+	}
+
+	return func(ctx context.Context, pushReq *push.Request) (*mimirpb.WriteResponse, error) {
+		userID, err := tenant.TenantID(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		req, err := pushReq.WriteRequest()
+		if err != nil {
+			return nil, err
+		}
+
+		// No ephemeral series from ruler.
+		if req.Source == mimirpb.RULE {
+			return next(ctx, pushReq)
+		}
+
+		req.EphemeralTimeseries = nil
+
+		ephemeralChecker := d.EphemeralCheckerByUser.EphemeralChecker(userID)
+		if ephemeralChecker != nil {
+			req.EphemeralTimeseries = mimirpb.PreallocTimeseriesSliceFromPool()
```

Comment:
Shall we get the slice only if there is actually an ephemeral series?

Reply:
good catch, thx.
done here 7e0caf58b